src/pow.rs

```diff
@@ -0,0 +1,1892 @@
+    ///
+    /// For example, a difficulty of 6,695,826 means that at a given hash rate, it will, on average,
+    /// take ~6.6 million times as long to find a valid block as it would at a difficulty of 1, or
+    /// alternatively, it will take, again on average, ~6.6 million times as many hashes to find a
+    /// valid block
+    ///
+    /// # Note
+    ///
+    /// Difficulty is calculated using the following algorithm `max / current` where [max] is
+    /// defined for the Bitcoin network and `current` is the current [target] for this block. As
+    /// such, a low target implies a high difficulty. Since [`Target`] is represented as a 256 bit
+    /// integer but `difficulty()` returns only 128 bits this means for targets below approximately
+    /// `0xffff_ffff_ffff_ffff_ffff_ffff` `difficulty()` will saturate at `u128::MAX`.
+    ///
+    /// [max]: crate::blockdata::constants::max_target
+    /// [target]: crate::blockdata::block::BlockHeader::target
+    pub fn difficulty(&self) -> u128 {
+        const MAX: Uint256 = Uint256([0, u128::max_value()]);
+
+        let d = constants::max_target().0 / self.0;
+
+        if d > MAX {
+            u128::max_value()
+        } else {
+            d.low_u128()
+        }
+    }
+}
+do_impl!(Target);
+
+impl Div for Target {
+    type Output = Self;
+    fn div(self, rhs: Self) -> Self { Self(self.0 / rhs.0) }
+}
+
+/// Converts $arr ([u128; 2]) to a big-endian array in $out.
+macro_rules! to_big_endian {
+    ($arr:expr, $out:expr, $half:literal, $shift_mod:literal, $shift:literal, $mask:literal, $type:ident) => {
+        for (i, src) in $arr.iter().enumerate() {
+            let start = i * $half;
+            for (j, dst) in $out[start..start + $half].iter_mut().enumerate() {
+                *dst = (src >> ($shift - j * $shift_mod) & $mask) as $type;
+            }
+        }
+    };
+}
+
+/// Converts $arr ([u128; 2]) to a little-endian array in $out.
+macro_rules! to_little_endian {
+    ($arr:expr, $out:expr, $half:literal, $shift_mod:literal, $shift:literal, $mask:literal, $type:ident) => {
+        for (i, src) in $arr.iter().rev().enumerate() {
+            let start = i * $half;
+            for (j, dst) in $out[start..start + $half].iter_mut().rev().enumerate() {
+                *dst = (src >> ($shift - j * $shift_mod) & $mask) as $type;
+            }
+        }
+    };
+}
+
+/// Big-endian 256 bit integer type.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+struct Uint256(pub [u128; 2]);
+
+impl Uint256 {
+    const MAX: Uint256 = Uint256([
+        0xffff_ffff_ffff_ffff_ffff_ffff_ffff_ffff,
+        0xffff_ffff_ffff_ffff_ffff_ffff_ffff_ffff,
+    ]);
+
+    const ZERO: Uint256 = Uint256([0, 0]);
+
+    const ONE: Uint256 = Uint256([0, 1]);
+
+    /// Creates `Self` from a big-endian array of `u128`s.
+    fn from_be_array_u128(a: [u128; 2]) -> Uint256 { Uint256(a) }
```

Comment:
This is cool but why do we have a gazillion different conversion functions for different integer types? Why not just bytes? (And even then it's not totally obvious what the purpose is...serialization I guess).